### PR TITLE
postprocessing

### LINKED ISSUE #172: mpi4py installation
I have been struggling with `mpi4py` installation in #164. I followed all the [instructions here](https://jonescompneurolab.github.io/hnn-core/parallel.html) and ran the following command: 

```py
$ mpiexec -np 2 nrniv -mpi -python -c 'from neuron import h; from mpi4py import MPI; \
                                       print("Hello from proc %d" % MPI.COMM_WORLD.Get_rank()); \
                                           h.quit()'
```

which runs as expected and gives:

```py
numprocs=2
NEURON -- VERSION 7.8.1-1-gda13bb7c HEAD (da13bb7c) 2020-07-22
Duke, Yale, and the BlueBrain Project -- Copyright 1984-2019
See http://neuron.yale.edu/neuron/credits

Hello from proc 0
Hello from proc 1
```

However, when I run the test:

```py
$ pytest ./hnn_core/tests/test_compare_hnn.py::test_mpi --pdb
```

I get:

```sh
hnn_core/tests/test_compare_hnn.py::test_mpi
  /Users/mainak/Documents/github_repos/hnn-core/hnn_core/parallel_backends.py:186: UserWarning: mpi4py not installed. will run on single processor
    warn('mpi4py not installed. will run on single processor')
```

Also I think when `mpi4py` is not installed, the test [should be skipped](https://docs.pytest.org/en/latest/skipping.html?highlight=skipping#skipping-test-functions) instead of passing.

### LINKED ISSUE #183: access raw dipole for each trial
The raw dipole before smoothing should be accessible for inspection and saving.

Perhaps `simulate` should return a rawdpl instead of the smoothed and scaled version.

https://github.com/jonescompneurolab/hnn-core/blob/073ef6c6a184677d4349e5812f1ed845857f6f74/hnn_core/dipole.py#L48

Then these lines can be moved to `simulate_dipole` after `_gather_trial_data`:
https://github.com/jonescompneurolab/hnn-core/blob/073ef6c6a184677d4349e5812f1ed845857f6f74/hnn_core/network_builder.py#L120-L123

Saving is probably best done with a parameter to `simulate_dipole` that will call `dpl.write()` before the lines above.

The parameter `save_dpl` can then be removed (used here):
https://github.com/jonescompneurolab/hnn-core/blob/073ef6c6a184677d4349e5812f1ed845857f6f74/hnn_core/network_builder.py#L117-L118

### PR DESCRIPTION
**Pull request by kohl-carmen**
Closes #183 
closes #172

### CODE REVIEW

**Comment by jasmainak** [comment]
@blakecaldwell do you mind taking a look to see if it addresses what you had in mind for #183 ?

@kohl-carmen one thing that remains to be checked is what does the GUI output as `rawdpl.txt`, if it's baseline renormalized or not?

**Comment by jasmainak** [comment]
another thing to do here is to add a test for the new `postproc` parameter.

File: hnn_core/pyramidal.py
Comment (jasmainak) [inline hnn_core/pyramidal.py:22]: changes are unrelated to the PR, so best to undo them as we discussed!

**Comment by blakecaldwell** [comment]
Just getting around to reviewing this. Sorry for the delay. Yes, this is looking very good and should resolve #183 with some minor updates. We briefly touched on this on Monday, but I have some specific suggestions now.

Since HNN GUI needs to write both `rawdpl.txt` and `dpl.txt`, I think this could be done by making a `postproc()` member of `Dipole` that gets called in `simulate` when `post_proc=True`. This would allow HNN GUI to call  `simulate_dipole(..., post_proc=False)`, write `rawdpl.txt`, and then `dpl.postproc(params)` before writing the `dpl.txt` file.

```
def postproc(self, params):
    dpl.baseline_renormalize(params)
    dpl.convert_fAm_to_nAm()
    dpl.scale(params['dipole_scalefctr'])
    dpl.smooth(params['dipole_smooth_win'] / params['dt'])
```

Note, the relevant params could be passed instead of the dictionary, perhaps preferable? I didn't look up what params `baseline_renormalize` needs.

File: hnn_core/network_builder.py
Comment (blakecaldwell) [inline hnn_core/network_builder.py:122]:
```
        if postproc:
            self.postproc(params)
```

**Comment by jasmainak** [comment]
I see, I would make it `_postproc` so it's a private method used only in the GUI. Don't want to encourage the pattern of users calling `foo_function(params)` from their Python scripts. Alternatively, if it makes sense, as you suggested, we can have explicitly the parameters for all the postprocessing methods ...

@kohl-carmen I saw you pushed some new commits. Let us know when we should take a look again. I get notifications only for comments, not for commits.

**Comment by blakecaldwell** [comment]
I agree with using specific parameters. It somewhat negates the point below, but I think it's worth pointing out.

I blanket disagree with making functions HNN use private. HNN should be considered a first-class use case of hnn-core.

**Comment by jasmainak** [comment]
Sure, of course HNN is a first class citizen of hnn-core. However, the reason I suggested using a private method is that it gives us flexibility to change the internals. If users start using hnn-core by passing params, they will complain if we try to change it in the future.

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:203]:
```suggestion
            Number of Pyramidal cells in x direction
```

**Comment by jasmainak** [comment]
@kohl-carmen can you update a test?

File: hnn_core/parallel_backends.py
Comment (jasmainak) [inline hnn_core/parallel_backends.py:257]: but what if `n_procs > 1` ? Will `postproc` work for MPI backend?

File: hnn_core/tests/test_dipole.py
Comment (rythorpe) [inline hnn_core/tests/test_dipole.py:61]: If we need to run simulations for this unit test, would it be best to put it in test_parallel_backends.py?

File: hnn_core/tests/test_dipole.py
Comment (rythorpe) [inline hnn_core/tests/test_dipole.py:63]: It might be a good idea to explicitly set the relative (`rtol`) and absolute (`atol`) tolerance args for [`assert_allclose()`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html) so that we can control how much error is acceptable. The defaults are `rtol=1e-07` and `atol=0`.

File: hnn_core/tests/test_dipole.py
Comment (rythorpe) [inline hnn_core/tests/test_dipole.py:67]: Same as [above](https://github.com/jonescompneurolab/hnn-core/pull/188/files#r523244077).

**Comment by kohl-carmen** [comment]
@klankinen and I tried to fix the postproc test. We set the std of the inputs to 0, rather than changing the tolerance on assert_allclose() because we didn't know what a reasonable tolerance would be?

The postproc test is currently still in test_dipole.py but we're happy to move it (might need some help)

**Comment by rythorpe** [comment]
> @klankinen and I tried to fix the postproc test. We set the std of the inputs to 0, rather than changing the tolerance on assert_allclose() because we didn't know what a reasonable tolerance would be?

Hmm, setting the std of the input times to 0 won't control for error induced by smoothing vs. not smoothing a given trial. My recommendation would be leave jitter in-between trials (i.e., reset start time std values to default) and solve for a reasonable tolerance as follows:

1. Set the smoothing kernel to be super narrow such that it only averages a few time samples surrounding the central time sample.

2. Given that that smoothing kernel we use is a hamming window, the maximal difference between the original central time sample and the smoothed version cannot exceed the difference between the original central time sample and the mean of all samples contained in the smoothing window.

3. Use (2) to compute the maximal relative difference (i.e., tolerance) between the post-processed time series and the original time series.

This method might be overkill but it should only take a few lines of code and can serve as a sanity check for any subsequent changes made to the post-processing steps in later PRs.

> The postproc test is currently still in test_dipole.py but we're happy to move it (might need some help)

Idk, @jasmainak @ntolley what do you think?

**Comment by jasmainak** [comment]
If you guys are working together you can both get credit for commits by [co-authoring them](https://docs.github.com/en/free-pro-team@latest/github/committing-changes-to-your-project/creating-a-commit-with-multiple-authors). Just wanted to put it out there for future commits. You can also [edit older commits and add authors](https://stackoverflow.com/questions/3042437/how-to-change-the-commit-author-for-one-specific-commit) but probably not worth the effort unless @klankinen insists ;-)

**Comment by jasmainak** [comment]
> because we didn't know what a reasonable tolerance would be?

Something reasonably small ... Is there a reason they are not exactly equal? Usually the defaults should do but if not [you can look around in the codebase](https://github.com/jonescompneurolab/hnn-core/blob/56f3565dcee8592872ac79a911b250c830d1d67b/hnn_core/tests/test_parallel_backends.py#L95) for tolerance used.

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:189]: lovely!

File: hnn_core/dipole.py
Comment (jasmainak) [inline hnn_core/dipole.py:151]: public method is missing docstring

File: hnn_core/parallel_backends.py
Comment (jasmainak) [inline hnn_core/parallel_backends.py:134]: do you think we could avoid the duplication of code by putting this inside `_gather_trial_data` with `postproc` as an argument?

**Comment by jasmainak** [comment]
> The postproc test is currently still in test_dipole.py but we're happy to move it (might need some help)

ideally, we should avoid duplication but let's get the other stuff working

also you need to rebase. Do you need help? I think @ntolley is an expert ;-)

last thing is you need to update `whats_new.rst`. Do it last to avoid having to rebase again (the file gets updated in almost every PR).

File: hnn_core/tests/test_dipole.py
Comment (jasmainak) [reply hnn_core/tests/test_dipole.py:66]: Defaults are usually pretty stringent. Unless there is a specific reason to have more loose checks ...

**Comment by ntolley** [comment]
On first glance it looks like the rebase should be pretty easy, I think most of the conflicts can be resolved by accepting both changes (one exception is adding the new argument to `simulate_dipole()`). Feel free to ping me if you have any questions.

**Comment by rythorpe** [comment]
> Something reasonably small ... Is there a reason they are not exactly equal? Usually the defaults should do but if not [you can look around in the codebase](https://github.com/jonescompneurolab/hnn-core/blob/56f3565dcee8592872ac79a911b250c830d1d67b/hnn_core/tests/test_parallel_backends.py#L95) for tolerance used.

This probably won't work because the error [here](https://github.com/jonescompneurolab/hnn-core/blob/56f3565dcee8592872ac79a911b250c830d1d67b/hnn_core/tests/test_parallel_backends.py#L95) is induced by rounding error rather than by convolution with a smoothing kernel.

File: hnn_core/tests/test_dipole.py
Comment (rythorpe) [reply hnn_core/tests/test_dipole.py:66]: See https://github.com/jonescompneurolab/hnn-core/pull/188#issuecomment-730815655.

**Comment by jasmainak** [comment]
> rather than by convolution with a smoothing kernel.

not sure I follow. There are two tests: 1) checking if `dpl_raw` and `dpl` are _different_. and 2) checking if `dpl_raw.post_proc(...)` gives the same result as `simulate_dipole(..., post_proc=True)`. It's doing the same arithmetic unless I'm missing something?

**Comment by rythorpe** [comment]
> not sure I follow. There are two tests: 1) checking if `dpl_raw` and `dpl` are _different_. and 2) checking if `dpl_raw.post_proc(...)` gives the same result as `simulate_dipole(..., post_proc=True)`. It's doing the same arithmetic unless I'm missing something?

Oops, somehow I missed test_dipole.py L67. For some reason I thought the goal was to simultaneously test the scope of the post-processing.

File: hnn_core/dipole.py
Comment (ntolley) [inline hnn_core/dipole.py:20]: This is one of the lines with a merge conflict due to the somatic currents PR. I'd recommend changing it to:
```py
def simulate_dipole(net, n_trials=None, postproc=True, record_vsoma=False, record_isoma=False):
```

File: hnn_core/dipole.py
Comment (ntolley) [inline hnn_core/dipole.py:33]: Here is the other conflicting spot. I think post-processing should be listed first so the docstring would become
```
postproc : bool
     If False, no postprocessing applied to the dipole
record_vsoma : bool
     Option to record somatic voltages from cells
record_isoma : bool
     Option to record somatic currents from cells
```

File: hnn_core/tests/test_dipole.py
Comment (jasmainak) [inline hnn_core/tests/test_dipole.py:134]: what is this guy doing in there ? copy-paste issue?

**Comment by jasmainak** [comment]
@kohl-carmen can you update `whats_new.rst` and set the PR title to [MRG] when it's all ready from your end for merging (pending reviews from others)?

File: hnn_core/tests/test_dipole.py
Comment (kohl-carmen) [reply hnn_core/tests/test_dipole.py:134]: I put that in there because a few lines down, we're comparing joblib and mpi. I assume this could be done better (taking joblib_net from the previous function?), but I didn't know how.

File: hnn_core/dipole.py
Comment (jasmainak) [reply hnn_core/dipole.py:36]: I'm fine with `postproc` being listed last as long as it matches the function signature. You have less chance of breaking existing user code when you add an argument to the end of the function. E.g., if someone did:

```py
record_vsoma = True
simulate_dipole(net, n_trials, record_vsoma, record_isoma)
```

now if you add `postproc` in the middle, it will get passed as `record_vsoma` when the user intended otherwise

File: hnn_core/tests/conftest.py
Comment (jasmainak) [inline hnn_core/tests/conftest.py:76]: @kohl-carmen if you add a `postproc` parameter here, then you can avoid duplication of code

**Review by jasmainak — APPROVED** [review]
PR is good for me now! I snuck in a couple of fixes related to requires_mpi4py and moving run_hnn_core into a pytest fixture.

@ntolley @blakecaldwell @rythorpe if you guys are happy please feel free to merge and/or leave comments if something is to be fixed

File: hnn_core/tests/test_dipole.py
Comment (blakecaldwell) [inline hnn_core/tests/test_dipole.py:85]: Nice, I didn't even realize we had a MPIBackend test in another file besides parallel_backends.py. Thanks @jasmainak for making the testing changes.

File: hnn_core/tests/test_dipole.py
Comment (blakecaldwell) [inline hnn_core/tests/test_dipole.py:93]: The "n_jobs" parameter doesn't do anything for MPIBackend. It defaults to running JoblibBackend. What about setting "n_procs=2"?

For a future PR that will help avoid this, I would suggest using "parallel" to replace "n_jobs" and "n_procs".

File: hnn_core/tests/conftest.py
Comment (blakecaldwell) [inline hnn_core/tests/conftest.py:74]: what about putting "fixture" in this function name? Otherwise it looks like a mysterious parameter to other testing functions.

File: hnn_core/dipole.py
Comment (blakecaldwell) [reply hnn_core/dipole.py:235]: add this change in next commit?

**Review by blakecaldwell — CHANGES_REQUESTED** [review]
Just have a few suggestions, mainly in the new testing code added. So close!

File: hnn_core/tests/conftest.py
Comment (jasmainak) [reply hnn_core/tests/conftest.py:74]: +1 @kohl-carmen can you make the changes yourself?

**Review by blakecaldwell — APPROVED** [review]
This looks great! Thanks @kohl-carmen

**Comment by blakecaldwell** [comment]
@jasmainak would you do a "squash and merge" or an automatic rebase to merge this PR? Not sure what your preference is.

**Comment by jasmainak** [comment]
I would do a "rebase and merge". Would give @kohl-carmen the full credit for the effort over the past month. Once we have establish set of contributors, we can move to "squash and merge" model

File: hnn_core/tests/test_dipole.py
Comment (ntolley) [reply hnn_core/tests/test_dipole.py:86]: Ah those would be my changes, thanks for fixing my ad-hoc MPI tests @jasmainak.

**Pull request merged**
